### PR TITLE
ENG-2914: Introduce GPUUsageMax

### PR DESCRIPTION
**Pull request by kubecost-pb**
## What does this PR change?
* Introduce GPUUsageMax to be collected from prometheus

## Does this PR relate to any other PRs?
https://kubecost.atlassian.net/browse/ENG-2959

## How will this PR impact users?
* 

## Does this PR address any GitHub or Zendesk issues?
N/A

## How was this PR tested?
* 

## Does this PR require changes to documentation?
* 

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
*

### CODE REVIEW

**Comment by avrodrigues5** [comment]
@kubecost-pb Other max reside in https://github.com/opencost/opencost/blob/develop/core/pkg/opencost/allocation.go#L92 rawAllocationOnly inside the allocation struct. I believe thats the best place for this to reside.

**Comment by cliffcolvin** [comment]
@kubecost-pb can you sign off on your commits? example below:
```
To add your Signed-off-by line to every commit in this branch:

Ensure you have a local copy of your branch by [checking out the pull request locally via command line](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally).
In your local branch, run: git rebase HEAD~2 --signoff
Force push your changes to overwrite the branch: git push --force-with-lease origin ENG-2914
```

you can also setup to have this automatically on all commits if you wish.

File: pkg/costmodel/allocation.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ const (
 	queryFmtCPUUsageAvg                 = `avg(rate(container_cpu_usage_seconds_total{container!="", container_name!="POD", container!="POD", %s}[%s])) by (container_name, container, pod_name, pod, namespace, instance, %s)`
 	queryFmtGPUsRequested               = `avg(avg_over_time(kube_pod_container_resource_requests{resource="nvidia_com_gpu", container!="",container!="POD", node!="", %s}[%s])) by (container, pod, namespace, node, %s)`
 	queryFmtGPUsUsageAvg                = `avg(avg_over_time(DCGM_FI_PROF_GR_ENGINE_ACTIVE{container!=""}[%s])) by (container, pod, namespace, %s)`
+	queryFmtGPUsUsageMax                = `max(max_over_time(DCGM_FI_PROF_GR_ENGINE_ACTIVE{container!=""}[%s])) by (container, pod, namespace, %s)`
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation.go:32]:
```suggestion
	queryFmtGPUsUsageMax                = `avg(max_over_time(DCGM_FI_PROF_GR_ENGINE_ACTIVE{container!=""}[%s])) by (container, pod, namespace, %s)`
```

Fact that we are doing average of all average Usage of GPU which is incorrect. I believe the best approach would be to average out all max. WDYT?

File: pkg/costmodel/allocation.go [hunk 2]
```diff
@@ -615,7 +620,8 @@ func (cm *CostModel) computeAllocation(start, end time.Time, resolution time.Dur
 	applyRAMBytesRequested(podMap, resRAMRequests, podUIDKeyMap)
 	applyRAMBytesUsedAvg(podMap, resRAMUsageAvg, podUIDKeyMap)
 	applyRAMBytesUsedMax(podMap, resRAMUsageMax, podUIDKeyMap)
-	applyGPUUsageAvg(podMap, resGPUsUsageAvg, podUIDKeyMap)
+	applyGPUUsage(podMap, resGPUsUsageAvg, podUIDKeyMap, GpuUsageConfig{Mode: "Average"})
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation.go]: This is good, but let's make average and max a constant in core/pkg/opencost. core/pkg/opencost is the most deep package. If you are declaring constants and GPUUsageStruct, it is better to declare them here in allocation.go than in allocation_helpers.go... 

I dont feel there is a need for GPUUsageConfig struct with just mode string.

File: pkg/costmodel/allocation_helpers.go [hunk 3]
```diff
@@ -642,15 +647,21 @@ func applyGPUUsageAvg(podMap map[podKey]*pod, resGPUUsageAvg []*prom.QueryResult
...
 				continue
 			}
 			if _, ok := thisPod.Allocations[container]; !ok {
 				thisPod.appendContainer(container)
 			}
 
 			// DCGM_FI_PROF_GR_ENGINE_ACTIVE metric is a float between 0-1.
-			thisPod.Allocations[container].GPUUsageAverage = res.Values[0].Value
+			for _, option := range opts {
+				if option.Mode == "Average" {
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation_helpers.go]: constants plz and also declare them inside core/pkg/opencost and refer here!

**Comment by thomasvn** [comment]
Agree with @avrodrigues5 that the best place to add the new field `GPUUsageMax` is probably in [RawAllocationOnlyData](https://github.com/opencost/opencost/blob/016dc897bf282640bc5382323c0f32fc25837358/core/pkg/opencost/allocation.go#L152-L179). Just for consistency?

**Comment by avrodrigues5** [comment]
@kubecost-pb  Could u sign off commits as mentioned by Cliff above?

Add some testing notes too.

@mbolt35 This looks ok to me, I would appreciate you giving a second review on this pretty, please.

**Comment by kubecost-pb** [comment]
> @kubecost-pb Could u sign off commits as mentioned by Cliff above?

It is signed. Do you not see it?

**Comment by avrodrigues5** [comment]
@kubecost-pb  only last two commits are signed off not the first ... easiest way to check that would be  having DCO checked marked. 

Screenshot 2024-10-09 at 11 00 25 AM

File: core/pkg/opencost/allocation.go [hunk 4]
```diff
@@ -174,8 +229,9 @@ func (lba *LbAllocation) SanitizeNaN() {
 // then this type would be unnecessary and its fields would go into the regular Allocation
 // and not in the AggregatedAllocation.
 type RawAllocationOnlyData struct {
-	CPUCoreUsageMax  float64 `json:"cpuCoreUsageMax"`
-	RAMBytesUsageMax float64 `json:"ramByteUsageMax"`
+	CPUCoreUsageMax  float64  `json:"cpuCoreUsageMax"`
+	RAMBytesUsageMax float64  `json:"ramByteUsageMax"`
+	GPUUsageMax      *float64 `json:"gpuUsageMax"` //@bingen:field[version=23]
```
Comment (avrodrigues5) [inline core/pkg/opencost/allocation.go:234]: @mbolt35 does adding field in this position pose any issues?

File: pkg/costmodel/allocation.go [hunk 5]
```diff
@@ -66,6 +67,8 @@ const (
 	queryFmtLBActiveMins                = `count(kubecost_load_balancer_cost{%s}) by (namespace, service_name, %s)[%s:%s]`
 	queryFmtOldestSample                = `min_over_time(timestamp(group(node_cpu_hourly_cost{%s}))[%s:%s])`
 	queryFmtNewestSample                = `max_over_time(timestamp(group(node_cpu_hourly_cost{%s}))[%s:%s])`
+	queryFmtIsGPuShared                 = `avg(avg_over_time(kube_pod_container_resource_requests{container!="", node != "", pod != "", container!= "", unit = "integer",  %s}[%s])) by (container, pod, namespace, node, resource)`
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation.go:70]:
```suggestion
	queryFmtIsGPUShared                 = `avg(avg_over_time(kube_pod_container_resource_requests{container!="", node != "", pod != "", container!= "", unit = "integer",  %s}[%s])) by (container, pod, namespace, node, resource)`
```

File: pkg/costmodel/allocation.go [hunk 6]
```diff
@@ -66,6 +67,8 @@ const (
...
+	queryFmtGetGPuInfo                  = `avg(avg_over_time(DCGM_FI_DEV_DEC_UTIL{container!="",%s}[%s])) by (container, pod, namespace, device, modelName, UUID)`
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation.go:71]:
```suggestion
	queryFmtGetGPUInfo                  = `avg(avg_over_time(DCGM_FI_DEV_DEC_UTIL{container!="",%s}[%s])) by (container, pod, namespace, device, modelName, UUID)`
```

File: pkg/costmodel/allocation.go [hunk 7]
```diff
@@ -426,12 +433,17 @@ func (cm *CostModel) computeAllocation(start, end time.Time, resolution time.Dur
 		}
 	}
 
+	// GPU Queries
+	//queryIsGpuShared := fmt.Sprintf(queryFmtIsGPuShared, durStr)
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation.go:437]: have to remove?

File: pkg/costmodel/allocation.go [hunk 8]
```diff
@@ -492,6 +504,13 @@ func (cm *CostModel) computeAllocation(start, end time.Time, resolution time.Dur
 	queryNetInternetCostPerGiB := fmt.Sprintf(queryFmtNetInternetCostPerGiB, env.GetPromClusterFilter(), durStr, env.GetPromClusterLabel())
 	resChNetInternetCostPerGiB := ctx.QueryAtTime(queryNetInternetCostPerGiB, end)
 
+	//GPU Queries
+	queryIsGpuShared := fmt.Sprintf(queryFmtIsGPuShared, env.GetPromClusterFilter(), durStr)
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation.go:508]:
```suggestion
	queryIsGPUShared := fmt.Sprintf(queryFmtIsGPuShared, env.GetPromClusterFilter(), durStr)
```

File: pkg/costmodel/allocation.go [hunk 9]
```diff
@@ -492,6 +504,13 @@ func (cm *CostModel) computeAllocation(start, end time.Time, resolution time.Dur
...
+	resChIsGpuShared := ctx.QueryAtTime(queryIsGpuShared, end)
```
Comment (avrodrigues5) [inline pkg/costmodel/allocation.go:509]:
```suggestion
	resChIsGPUShared := ctx.QueryAtTime(queryIsGpuShared, end)
```

File: core/pkg/opencost/allocation.go (hunk 4 above)
Comment (mbolt35) [reply core/pkg/opencost/allocation.go:234]: Nope! This looks fine :)

**Pull request closed without merging**